### PR TITLE
Fixing TokenStreamLintRule crashes on ternary expression

### PR DESCRIPTION
**Pull request by pawelsag**
Fixing #689

### CODE REVIEW

File: verilog/analysis/checkers/token_stream_lint_rule.cc [hunk 1]
```diff
@@ -69,7 +70,16 @@ void TokenStreamLintRule::HandleSymbol(const verible::Symbol& symbol,
   verible::matcher::BoundSymbolManager manager;
   if (StringLiteralMatcher().Matches(symbol, &manager)) {
     const auto& string_node = SymbolCastToNode(symbol);
-    const auto& string_literal = SymbolCastToLeaf(*string_node.children()[0]);
+    const auto& node_children = string_node.children();
+    const auto& literal =
```
Comment (hzeller) [inline verilog/analysis/checkers/token_stream_lint_rule.cc:74]: So could it be that we actually can have more than one string literal in this node ?
I suspect in our ternary situation we might have some TK_StringLiteral ':' TK_StringLiteral ?
Here, we're only looking for the first occurence.

So if there are multiple, should we check all of the children's string-literals ?
(might actually simplify the code as we can simply loop through the node-children and `continue` on non-`TK_StringLiteral`-ness).

File: verilog/analysis/checkers/token_stream_lint_rule.cc [hunk 2]
```diff
@@ -69,7 +70,16 @@ void TokenStreamLintRule::HandleSymbol(const verible::Symbol& symbol,
   verible::matcher::BoundSymbolManager manager;
   if (StringLiteralMatcher().Matches(symbol, &manager)) {
```
Comment (hzeller) [inline verilog/analysis/checkers/token_stream_lint_rule.cc:71]: To keep the indentation level flat, we can formulate this first test with an early return
```
  if (!StringLiteralMatcher().Matches(symbol, &manager))
    return;
```

File: verilog/analysis/checkers/token_stream_lint_rule_test.cc [hunk 3]
```diff
@@ -41,6 +41,7 @@ TEST(StringLiteralConcatenationTest, FunctionFailures) {
        "\"Humpty Dumpty had a great fall.\"};", "\nendmodule"},
       {"module m;\n", "string tmp = {\"Humpty Dumpty \\ sat on a wall.\",",
        "\"Humpty Dumpty had a great fall.\"};", "\nendmodule"},
+      {"module a();\nstring x = 1 ? \"foo\" : \"bar\";\nendmodule"},
```
Comment (hzeller) [inline verilog/analysis/checkers/token_stream_lint_rule_test.cc:44]: Are there otherr situations where the previous code would've stumbled upon ? If you are aware of some, we should probably add them.

Also since we now know that `?:` is actually showing up as more complicated node-structure, we should test that we can detect line-continuation attempts in each branch (line continuation attempted in`"foo"` or in `"bar"` ?)

**Review by hzeller — COMMENTED** [review]
Some suggestions for additional tests and possibly better coverage of corner-cases.

File: verilog/analysis/checkers/token_stream_lint_rule.cc (hunk 1 above)
Comment (pawelsag) [reply verilog/analysis/checkers/token_stream_lint_rule.cc:74]: In our ternary example when I printed the tree I found such structure:
```
Node @0 (tag: kConditionExpression) {
   Node @0 (tag: kNumber) {
      Leaf @0 (#TK_DecNumber @27-28: "1")
   }
   Leaf @1 (#'?' @29-30: "?")
   Node @2 (tag: kExpression) {
         Leaf @0 (#TK_StringLiteral @31-36: ""foo"")
   }
   Leaf @3 (#':' @37-38: ":")
   Leaf @4 (#TK_StringLiteral @39-44: ""bar"")
}
```
And as we can see we have two string literals. So the symbols passed to checker are: `kConditionExpression` and `kExpression`.  So for the first node, I'm looking for  `TK_StringLiteral` within its children, and the same is done for the second obtained symbol.
In the previous solution, I assumed that the string literal will be always in the first position in the children's leaf-list.

I believe that the later `if` validation can be removed cause the checker symbol is guaranteed to points to a node with TK_StringLiteral and there is no chance that there wouldn't be the desired string token.

So shouldn't I look for the string literal for each given symbol within the symbol's children? (but only leaves are considered, I do not dive into the nodes)
Do I understand it correctly?

File: verilog/analysis/checkers/token_stream_lint_rule_test.cc (hunk 3 above)
Comment (pawelsag) [reply verilog/analysis/checkers/token_stream_lint_rule_test.cc]: I have split the test structure, and few more test cases were added, but just for now I have no more ideas

File: verilog/analysis/checkers/token_stream_lint_rule_test.cc [hunk 4]
```diff
@@ -55,6 +67,15 @@ TEST(StringLiteralConcatenationTest, FunctionFailures) {
         "fall.\""},
        ";",
        "\nendmodule"},
+      {"module m;\nstring x = 1 ?",
+       {kToken, "\"foo\\\nincorrect\""},
+       " : \"bar\";\nendmodule"},
```
Comment (hzeller) [inline verilog/analysis/checkers/token_stream_lint_rule_test.cc:72]: Another test where the second branch of the ?:  `"\"bar\\\nis incorrect\""` maybe ?

**Review by hzeller — CHANGES_REQUESTED** [review]
Nice. One more test-case suggestion.

File: verilog/analysis/checkers/token_stream_lint_rule_test.cc (hunk 4 above)
Comment (pawelsag) [reply verilog/analysis/checkers/token_stream_lint_rule_test.cc:72]: Thanks for this test, I found one more bug when I applied the suggestion. The bug was related to incorrectly pointing token position.

**Review by hzeller — APPROVED** [review]
LGTM, Thanks!
